File: apps/web/lib/validations/template.ts
import * as z from "zod";

import { contentCreatorSchema, contentRole } from "@/lib/validations/content";

/** The three kinds shown on a template card (T1) — derived server-side from
 * `payload_kind`, not chosen by the author. */
export const templateKind = z.enum(["workflow", "dashboard", "layout"]);

/** What a template's frozen payload actually is (T1/T2): a workflow or
 * layout config snapshot, or a hidden frozen project copy. */
export const templatePayloadKind = z.enum(["workflow", "layout", "project"]);

/** The GOAT catalog shelf state (T4). v1 only ever moves between `none` and
 * `published`; `proposed`/`declined` are reserved for the later submission flow. */
export const templateCatalogStatus = z.enum(["none", "proposed", "published", "declined"]);

/** A declared input slot on a saved template (T5) — mirrors
 * `core.schemas.template.TemplateInput`. */
export const templateInputSchema = z.object({
  key: z.string(),
  label: z.string(),
  mode: z.enum(["ship", "ask"]),
  layer_id: z.string().uuid().nullable(),
  layer_type: z.enum(["feature", "table", "raster"]).nullable(),
  geometry_type: z.string().nullable(),
  from_catalog: z.boolean().default(false),
});

/** Where a template is saved from, or re-snapshotted from on refresh (T8) —
 * mirrors `core.schemas.template.TemplateSource`. */
export const templateSourceSchema = z.object({
  kind: z.enum(["workflow", "layout", "project"]),
  project_id: z.string().uuid(),
  workflow_id: z.string().uuid().nullable().optional(),
  layout_id: z.string().uuid().nullable().optional(),
});

/** One row of a template preview's `datasets_needing_share` (T6) — a shipped
 * dataset the destination space's audience cannot read yet. */
export const datasetShareLineSchema = z.object({
  layer_id: z.string().uuid(),
  name: z.string(),
  from_catalog: z.boolean(),
  current_audience: z.enum(["personal", "team", "organization"]),
});

/** Response of `POST /template/preview`: what saving a source into a folder
 * would detect, without writing anything. `.passthrough()` so a field added
 * on the backend later (it is under active development) does not break
 * parsing here. */
export const templatePreviewSchema = z
  .object({
    detected_inputs: z.array(templateInputSchema),
    kinds: z.array(z.string()),
    datasets_needing_share: z.array(datasetShareLineSchema).default([]),
  })
  .passthrough();

/** Body of `POST /template` — mirrors `TemplateCreate`. */
export const templateCreateSchema = z.object({
  name: z.string(),
  description: z.string().nullable().optional(),
  categories: z.array(z.string()).default([]),
  folder_id: z.string().uuid(),
  source: templateSourceSchema,
  inputs: z.array(templateInputSchema).default([]),
  share_datasets: z.array(z.string().uuid()).default([]),
  thumbnail_url: z.string().nullable().optional(),
  /** For a layout payload: the page it prints on, read off the config by the
   * client and stored verbatim. */
  page_size: z.string().nullable().optional(),
  page_orientation: z.enum(["portrait", "landscape"]).nullable().optional(),
});

/** Body of `PATCH /template/{id}`: metadata, plus the folder a Move puts the
 * template in — that folder has to be one in the template's own space, since
 * leaving the space is a transfer, not a move. */
export const templateUpdateSchema = z.object({
  name: z.string().optional(),
  description: z.string().nullable().optional(),
  categories: z.array(z.string()).optional(),
  thumbnail_url: z.string().nullable().optional(),
  folder_id: z.string().uuid().optional(),
  /** Re-labelled with the config: a refresh re-freezes a layout, which can
   * change the page it prints on. */
  page_size: z.string().nullable().optional(),
  page_orientation: z.enum(["portrait", "landscape"]).nullable().optional(),
});

/** One node of a workflow preview descriptor: where it sits on the canvas,
 * how big it is there, what it is called and which node type it is. */
export const templatePreviewNodeSchema = z.object({
  label: z.string().default(""),
  type: z.string().default(""),
  x: z.number(),
  y: z.number(),
  w: z.number(),
  h: z.number(),
  /** Which of the node type's icons the canvas shows: a dataset's geometry
   * (`point`/`line`/`polygon`/`table`), a tool's process id, `export_dataset`
   * for an export, `if` for a branch. Null for an annotation, which shows
   * none, and absent on a descriptor built before the field existed. */
  icon: z.string().nullable().optional(),
  /** The colour the node is painted in, for the one node type that carries
   * one: an annotation's own `data.backgroundColor`, as `#rgb`/`#rrggbb`.
   * Null for every other type, for an annotation with no colour of its own,
   * and on a descriptor built before the field existed — the drawing then
   * paints the note in the colour the canvas defaults to. */
  color: z.string().nullable().optional(),
  /** An annotation's own note markup, sanitised to the drawing's allow-list
   * (`sanitizeNoteHtml`). Only the client builds it — from the config the
   * author owns, for the pre-save preview and the PNG — so it is optional
   * and a descriptor without it draws an untexted note card. */
  html: z.string().nullable().optional(),
});

/** The camera a layout's map frame is frozen at — the centre and the zoom the
 * element's own `config.viewState` carries. */
export const templatePreviewViewStateSchema = z.object({
  latitude: z.number(),
  longitude: z.number(),
  zoom: z.number(),
});

/** What a layout element is styled with, as the layout stores it on the
 * element's own `style` — the enabled halves of `elementStyleSchema` in
 * `lib/validations/reportLayout.ts`, reduced to what a drawing paints. A
 * field is carried only where the element actually asks for it: a layout
 * frames and fills nothing by default. */
export const templatePreviewElementStyleSchema = z.object({
  /** The frame the element draws, `width` in page millimetres. Absent where
   * it draws none. */
  border: z.object({ color: z.string(), width: z.number() }).nullable().optional(),
  /** The ground the element is filled with. Absent where it has none. */
  background: z.object({ color: z.string(), opacity: z.number() }).nullable().optional(),
  /** The inset the element keeps its content in, in page millimetres. */
  padding: z.number().nullable().optional(),
  /** How opaque the element is as a whole. Absent where it is opaque. */
  opacity: z.number().nullable().optional(),
});

/** One element of a layout preview descriptor, in page units.
 *
 * Beyond the box, an element carries the little of its own content a drawing
 * can show: the three payload fields below. Only the client builds them —
 * from the config the author owns — so each is optional, and a descriptor
 * without them draws the placeholder glyph the element type reads as. */
export const templatePreviewElementSchema = z.object({
  type: z.string().default(""),
  x: z.number(),
  y: z.number(),
  width: z.number(),
  height: z.number(),
  /** A map frame's frozen camera, which a drawing fetches a static basemap
   * frame for. Absent on every other element type, and on a map frame that
   * names no camera. */
  viewState: templatePreviewViewStateSchema.nullable().optional(),
  /** A text block's own markup, sanitised to the drawing's allow-list
   * (`sanitizeNoteHtml`) — so a title is drawn as its real words. */
  html: z.string().nullable().optional(),
  /** The size that markup's first sized run declares, in the layout canvas's
   * own pixels: the base a drawing sets the text at, scaled into its frame. */
  fontSize: z.number().nullable().optional(),
  /** A legend's own title, as plain text. */
  title: z.string().nullable().optional(),
  /** What the element is framed and filled with. Absent where it asks for
   * neither, which is what a layout element does by default. */
  style: templatePreviewElementStyleSchema.nullable().optional(),
});

/** The structure of a payload, as the client builds it from a config it can
 * see (`descriptorFromWorkflowConfig` / `descriptorFromLayoutConfig`):
 * enough to draw the payload, never any of its data. Purely local — nothing
 * stores or answers it. It is what the save dialog's live preview draws and
 * what the PNG writers are handed. */
export const templatePreviewDescriptorSchema = z.discriminatedUnion("kind", [
  z.object({
    kind: z.literal("workflow"),
    nodes: z.array(templatePreviewNodeSchema).default([]),
    /** Index pairs into `nodes`. */
    edges: z.array(z.tuple([z.number(), z.number()])).default([]),
  }),
  z.object({
    kind: z.literal("layout"),
    orientation: z.enum(["portrait", "landscape"]).default("portrait"),
    page: z.object({ width: z.number(), height: z.number() }),
    elements: z.array(templatePreviewElementSchema).default([]),
  }),
  z.object({
    kind: z.literal("project"),
    layers: z.number().int().default(0),
  }),
]);

/** `GET /template/{id}` only: where the template was saved from, with the
 * names the edit dialog links and whether a refresh from there can work —
 * mirrors `core.schemas.template.TemplateSourceInfo`. */
export const templateSourceInfoSchema = z.object({
  kind: templatePayloadKind,
  project_id: z.string().uuid().nullable(),
  project_name: z.string().nullable(),
  workflow_id: z.string().uuid().nullable().optional(),
  workflow_name: z.string().nullable().optional(),
  layout_id: z.string().uuid().nullable().optional(),
  layout_name: z.string().nullable().optional(),
  available: z.boolean(),
});

/** Response shape for the template API's read routes — mirrors `TemplateRead`. */
export const templateReadSchema = z.object({
  /** Filled by `POST /template/{id}/refresh` when re-snapshotting now ships
   * datasets the template's space audience cannot read yet; empty otherwise. */
  datasets_needing_share: z.array(datasetShareLineSchema).default([]),
  id: z.string().uuid(),
  name: z.string(),
  description: z.string().nullable(),
  categories: z.array(z.string()).default([]),
  thumbnail_url: z.string().nullable(),
  /** The page a layout template prints on, as its author saved it — the two
   * values a card labels it with without reading the frozen config. Null for
   * a workflow or project payload, and on a template saved before the
   * fields existed. */
  page_size: z.string().nullable().optional(),
  page_orientation: z.enum(["portrait", "landscape"]).nullable().optional(),
  space_id: z.string().uuid(),
  folder_id: z.string().uuid(),
  created_by: contentCreatorSchema.nullable().optional(),
  payload_kind: templatePayloadKind,
  kinds: z.array(templateKind),
  inputs: z.array(templateInputSchema).default([]),
  ships_sample_data: z.boolean().default(false),
  catalog_status: templateCatalogStatus,
  source_ref: z.record(z.unknown()).default({}),
  source: templateSourceInfoSchema.nullable().optional(),
  my_role: contentRole,
  created_at: z.string(),
  updated_at: z.string(),
  /** Only populated on `?include_config=true` for an owner/editor caller. */
  config: z.record(z.unknown()).nullable().optional(),
});

/** Response of `GET /template`: one page of the caller's readable templates. */
export const templatePageSchema = z.object({
  items: z.array(templateReadSchema),
  total: z.number().int(),
});

/** One row of `GET /template/categories`: a category in use on the caller's
 * readable templates, and how many of them carry it. Grouped
 * case-insensitively server-side, so `name` is one spelling of it. */
export const templateCategoryFacetSchema = z.object({
  name: z.string(),
  count: z.number().int(),
});

/** Body of `POST /template/{id}/use` (T7) — mirrors `TemplateUseRequest`. */
export const templateUseRequestSchema = z.object({
  project_id: z.string().uuid().nullable().optional(),
  target_folder_id: z.string().uuid().nullable().optional(),
  name: z.string().nullable().optional(),
  bindings: z.record(z.string().uuid()).default({}),
});

/** Response of `POST /template/{id}/use`: what got created, what got linked,
 * and which inputs still need the caller's attention. */
export const templateUseResultSchema = z.object({
  project_id: z.string().uuid(),
  workflow_id: z.string().uuid().nullable().optional(),
  layout_id: z.string().uuid().nullable().optional(),
  added_layer_project_ids: z.array(z.number().int()).default([]),
  unresolved_inputs: z.array(templateInputSchema).default([]),
});

/** Body of `POST /template/{id}/grant` — mirrors `TemplateGrantCreate`. */
export const templateGrantCreateSchema = z.object({
  grantee_type: z.enum(["user", "team", "organization"]),
  grantee_id: z.string().uuid(),
  role: z.enum(["template-viewer", "template-editor"]),
});

/** One row of `GET /template/{id}/grant` — mirrors `TemplateGrantRead`. */
export const templateGrantReadSchema = z.object({
  id: z.string().uuid(),
  grantee_type: z.enum(["user", "team", "organization"]),
  grantee_id: z.string().uuid(),
  grantee_name: z.string(),
  role: z.enum(["template-viewer", "template-editor"]),
  granted_by: z.string().uuid().nullable(),
  created_at: z.string(),
});

export const templateGrantsResponseSchema = z.object({
  grants: z.array(templateGrantReadSchema),
});

/** The template browser's source segments (T7/T9) — `all` unions every
 * scope the caller has. */
export type TemplateSourceFilter = "all" | "goat" | "mine" | "team" | "org";

export type TemplateKind = z.infer<typeof templateKind>;
export type TemplatePayloadKind = z.infer<typeof templatePayloadKind>;
export type TemplateCatalogStatus = z.infer<typeof templateCatalogStatus>;
export type TemplateInput = z.infer<typeof templateInputSchema>;
export type TemplateSource = z.infer<typeof templateSourceSchema>;
export type TemplateSourceInfo = z.infer<typeof templateSourceInfoSchema>;
export type DatasetShareLine = z.infer<typeof datasetShareLineSchema>;
export type TemplatePreview = z.infer<typeof templatePreviewSchema>;
export type TemplatePreviewNode = z.infer<typeof templatePreviewNodeSchema>;
export type TemplatePreviewElement = z.infer<typeof templatePreviewElementSchema>;
export type TemplatePreviewViewState = z.infer<typeof templatePreviewViewStateSchema>;
export type TemplatePreviewElementStyle = z.infer<typeof templatePreviewElementStyleSchema>;
export type TemplatePreviewDescriptor = z.infer<typeof templatePreviewDescriptorSchema>;
/** The workflow branch of the descriptor on its own — what the geometry and
 * the snapshot renderer take. */
export type TemplateWorkflowPreview = Extract<TemplatePreviewDescriptor, { kind: "workflow" }>;
export type TemplateLayoutPreview = Extract<TemplatePreviewDescriptor, { kind: "layout" }>;
/** The input shape, not the parsed/defaulted one — a caller building a save
 * request should not have to spell out `categories`/`inputs`/`share_datasets`
 * when they mean the backend's own `[]` default for each. */
export type TemplateCreate = z.input<typeof templateCreateSchema>;
export type TemplateUpdate = z.infer<typeof templateUpdateSchema>;
export type TemplateRead = z.infer<typeof templateReadSchema>;
export type TemplatePage = z.infer<typeof templatePageSchema>;
export type TemplateCategoryFacet = z.infer<typeof templateCategoryFacetSchema>;
/** Input shape — `bindings` defaults to `{}` and should stay optional for a
 * caller with nothing to bind. */
export type TemplateUseRequest = z.input<typeof templateUseRequestSchema>;
export type TemplateUseResult = z.infer<typeof templateUseResultSchema>;
export type TemplateGrantCreate = z.infer<typeof templateGrantCreateSchema>;
export type TemplateGrantRead = z.infer<typeof templateGrantReadSchema>;
export type TemplateGrantsResponse = z.infer<typeof templateGrantsResponseSchema>;

File: apps/core/src/core/schemas/template.py
"""Template schemas: input declarations (T5) plus the save/list/read/update
request and response models for the template API (T1, T2, T4, T6, T8)."""

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from core.schemas.content import ContentCreator

_LAYER_TYPES = ("feature", "table", "raster")


def normalize_layer_type(value: Any) -> Any:
    """``value`` when it is one of the three layer types, else ``None``.

    Workflow configs are stored unvalidated, so a dataset node's
    ``layerType`` can hold any spelling — older nodes carry ``"layer"``,
    saved before the web narrowed the field to feature|table|raster. An
    unrecognised value means "type unknown" rather than an error, since
    nothing downstream can act on it either way. The comparison is exact:
    the web writes these three lowercase, so ``"FEATURE"`` is unknown too.
    """
    return value if value in _LAYER_TYPES else None


class DetectedInput(BaseModel):
    """A dataset-node reference found while scanning a workflow config.

    Produced by ``core.templates.snapshot.detect_workflow_inputs``. Detection
    only reads the config: it knows nothing about the author's ship/ask
    choice, or the layer's catalog provenance — the caller turns each one
    into a ``TemplateInput`` after deciding ``mode`` and looking up
    ``from_catalog`` from the database. ``layer_type`` holds ``None`` for a
    node whose stored ``layerType`` is not one of the three types.
    """

    key: str
    label: str
    layer_id: UUID
    project_layer_id: int | None = None
    layer_type: Literal["feature", "table", "raster"] | None = None
    geometry_type: str | None = None

    @field_validator("layer_type", mode="before")
    @classmethod
    def _normalize_layer_type(cls, value: Any) -> Any:
        return normalize_layer_type(value)


class TemplateInput(BaseModel):
    """A declared input slot on a saved template (T5).

    ``key`` is the stable id a snapshot uses to address the slot again later:
    ``"node:<dataset node id>"`` for a workflow dataset node, or
    ``"param:<node id>:<param>"`` for a tool-parameter slot (not produced by
    this module yet). ``mode`` is the author's choice at save time: "ship"
    keeps ``layer_id`` bound so the dataset resolves for every user who can
    read it (catalog datasets resolve for everyone through promote-on-use);
    "ask" clears ``layer_id`` and turns the reference into a named, typed
    slot the user fills in on use. ``layer_type`` holds ``None`` for an input
    whose type is not one of the three types.
    """

    key: str
    label: str
    mode: Literal["ship", "ask"]
    layer_id: UUID | None = None
    layer_type: Literal["feature", "table", "raster"] | None = None
    geometry_type: str | None = None
    from_catalog: bool = False

    @field_validator("layer_type", mode="before")
    @classmethod
    def _normalize_layer_type(cls, value: Any) -> Any:
        return normalize_layer_type(value)


class TemplateSource(BaseModel):
    """Where a template is saved from (T8), or re-snapshotted from (refresh).

    ``workflow_id``/``layout_id`` are required for their matching ``kind``
    and ignored otherwise; a ``project`` source needs only ``project_id``.
    """

    kind: Literal["workflow", "layout", "project"]
    project_id: UUID
    workflow_id: UUID | None = None
    layout_id: UUID | None = None


class TemplateSourceInfo(BaseModel):
    """Where a template was saved from, resolved for the caller: names for
    the edit dialog's links, and whether "Update from source" can work —
    the project must still be readable and the workflow/layout still exist.
    Only ``GET /template/{id}`` fills this in; list rows carry ``None``."""

    kind: Literal["workflow", "layout", "project"]
    project_id: UUID | None
    project_name: str | None
    workflow_id: UUID | None = None
    workflow_name: str | None = None
    layout_id: UUID | None = None
    layout_name: str | None = None
    available: bool


class TemplatePreviewRequest(BaseModel):
    """Body of ``POST /template/preview``: what would saving this source
    into ``folder_id`` detect and require."""

    source: TemplateSource
    folder_id: UUID


class DatasetShareLine(BaseModel):
    """One row of ``TemplatePreview.datasets_needing_share`` (T6): a shipped
    dataset the destination space's audience cannot read yet."""

    layer_id: UUID
    name: str
    from_catalog: bool
    current_audience: Literal["personal", "team", "organization"]


class TemplatePreview(BaseModel):
    """Response of ``POST /template/preview``: what saving ``source`` into
    ``folder_id`` would detect, without writing anything."""

    detected_inputs: list[TemplateInput]
    kinds: list[str]
    datasets_needing_share: list[DatasetShareLine]


class TemplateCreate(BaseModel):
    """Body of ``POST /template`` (T8): the author's save-dialog choices."""

    name: str
    description: str | None = None
    categories: list[str] = []
    folder_id: UUID
    source: TemplateSource
    inputs: list[TemplateInput] = []
    share_datasets: list[UUID] = []
    thumbnail_url: str | None = None
    page_size: str | None = Field(
        default=None,
        description="For a layout payload: the page the layout prints on "
        '("A4", "A3", "Letter", …), read off the config by the client and '
        "stored verbatim. It is what a card labels the template with for a "
        "reader who cannot see the frozen config.",
    )
    page_orientation: Literal["portrait", "landscape"] | None = None


class TemplateUpdate(BaseModel):
    """Body of ``PATCH /template/{id}``: metadata only — the payload itself
    only changes through ``refresh``."""

    model_config = ConfigDict(extra="forbid")

    name: str | None = None
    description: str | None = None
    categories: list[str] | None = None
    thumbnail_url: str | None = None
    page_size: str | None = None
    page_orientation: Literal["portrait", "landscape"] | None = None
    folder_id: UUID | None = Field(
        default=None,
        description="Move the template to this folder. None means unchanged; "
        "the target must belong to the template's own space (its root/home "
        "folder is a valid target).",
    )


class TemplateRead(BaseModel):
    """Response shape for the template API's read routes."""

    id: UUID
    name: str
    description: str | None
    categories: list[str]
    thumbnail_url: str | None
    page_size: str | None = Field(
        default=None,
        description="For a layout payload: the page the layout prints on, as "
        "the client stored it. Returned to every reader, so a card can label "
        "the template without the frozen config. None for a workflow or "
        "project payload.",
    )
    page_orientation: Literal["portrait", "landscape"] | None = None
    space_id: UUID
    folder_id: UUID
    created_by: ContentCreator | None
    payload_kind: Literal["workflow", "layout", "project"]
    kinds: list[str]
    inputs: list[TemplateInput]
    ships_sample_data: bool
    catalog_status: Literal["none", "proposed", "published", "declined"]
    source_ref: dict[str, Any]
    source: TemplateSourceInfo | None = None
    my_role: Literal["owner", "editor", "viewer"]
    created_at: datetime
    updated_at: datetime
    config: dict[str, Any] | None = Field(
        default=None,
        description="Only populated on `GET /template/{id}?include_config=true` "
        "for an owner/editor caller, and only for a workflow/layout payload "
        "(a project payload's config lives on the frozen source project).",
    )
    datasets_needing_share: list[DatasetShareLine] = Field(
        default_factory=list,
        description="Only populated by `POST /template/{id}/refresh`: shipped "
        "inputs the reconciled config still ships but that are not yet "
        "readable by the template's own space — refresh does not "
        "auto-grant these (T6), it only reports them.",
    )


class TemplatePage(BaseModel):
    """Response of ``GET /template``: one page of the caller's readable
    templates."""

    items: list[TemplateRead]
    total: int


class TemplateCategoryFacet(BaseModel):
    """One row of ``GET /template/categories``: a category in use on the
    caller's readable templates, and how many of them carry it.

    Categories are free-form strings grouped case-insensitively; ``name``
    is the spelling of the template that first used it (earliest
    ``created_at``), so "Mobility" and "mobility" are one row.
    """

    name: str
    count: int


class TemplateUseRequest(BaseModel):
    """Body of ``POST /template/{id}/use`` (T7): where to land the
    template's payload and how to fill its inputs.

    Exactly one of ``project_id`` (insert into that existing project —
    workflow/layout payloads only) or ``target_folder_id`` (create a new
    project there) is required; a project payload always creates a new
    project, so it requires ``target_folder_id``. ``bindings`` maps an
    input's ``key`` to a layer the caller chose — filling an "ask" slot,
    or overriding a "ship" input the caller cannot read (or simply wants
    to replace with their own dataset).
    """

    project_id: UUID | None = None
    target_folder_id: UUID | None = None
    name: str | None = None
    bindings: dict[str, UUID] = {}


class TemplateUseResult(BaseModel):
    """Response of ``POST /template/{id}/use``: what got created, what got
    linked, and which inputs still need the caller's attention."""

    project_id: UUID
    workflow_id: UUID | None = None
    layout_id: UUID | None = None
    added_layer_project_ids: list[int] = []
    unresolved_inputs: list[TemplateInput] = []


class TemplateGrantCreate(BaseModel):
    """Body of ``POST /template/{id}/grant`` (T3/T6): grant a role on a
    template to a user, team, or organization. Owner only."""

    grantee_type: Literal["user", "team", "organization"]
    grantee_id: UUID
    role: Literal["template-viewer", "template-editor"]


class TemplateGrantRead(BaseModel):
    """One row of ``GET /template/{id}/grant``: a direct grant on a template,
    ``id`` being the ``grant_id`` used by ``DELETE /template/{id}/grant/{grant_id}``."""

    id: UUID
    grantee_type: Literal["user", "team", "organization"]
    grantee_id: UUID
    grantee_name: str
    role: Literal["template-viewer", "template-editor"]
    granted_by: UUID | None
    created_at: datetime


class TemplateGrantsResponse(BaseModel):
    """Response of ``GET /template/{id}/grant``: every direct grant on the
    template."""

    grants: list[TemplateGrantRead]
